packages/react/src/utils/useScrollLock.ts

```diff
@@ -8,172 +9,104 @@ let originalBodyStyles = {};
-  };
+function supportsDvh() {
+  return (
+    typeof CSS !== 'undefined' &&
+    typeof CSS.supports === 'function' &&
+    CSS.supports('height', '1dvh')
+  );
 }
 
-function preventScrollIOS(referenceElement?: Element | null) {
+function hasInsetScrollbars(referenceElement?: Element | null) {
+  if (typeof document === 'undefined') {
+    return false;
+  }
   const doc = ownerDocument(referenceElement);
-  const html = doc.documentElement;
-  const body = doc.body;
-  const htmlStyle = html.style;
-  const bodyStyle = body.style;
-
-  // iOS 12 does not support `visualViewport`.
-  const { x, y } = getVisualOffsets(doc);
-  const scrollX = bodyStyle.left ? parseFloat(bodyStyle.left) : window.scrollX;
-  const scrollY = bodyStyle.top ? parseFloat(bodyStyle.top) : window.scrollY;
-
-  originalHtmlStyles = {
-    overflowX: htmlStyle.overflowX,
-    overflowY: htmlStyle.overflowY,
-  };
-
-  originalBodyStyles = {
-    position: bodyStyle.position,
-    top: bodyStyle.top,
-    left: bodyStyle.left,
-    right: bodyStyle.right,
-    overflowX: bodyStyle.overflowX,
-    overflowY: bodyStyle.overflowY,
-  };
-
-  Object.assign(htmlStyle, {
-    overflow: 'visible',
-  });
-
-  Object.assign(bodyStyle, {
-    position: 'fixed',
-    top: `${-(scrollY - y)}px`,
-    left: `${-(scrollX - x)}px`,
-    right: '0',
-    overflow: 'hidden',
-  });
-
-  return () => {
-    Object.assign(htmlStyle, originalHtmlStyles);
-    Object.assign(bodyStyle, originalBodyStyles);
-    window.scrollTo({ left: scrollX, top: scrollY, behavior: 'instant' });
-  };
+  const win = ownerWindow(doc);
+  return win.innerWidth - doc.documentElement.clientWidth > 0;
 }
 
 function preventScrollStandard(referenceElement?: Element | null) {
-  const isFirefox = /firefox/i.test(getUserAgent());
   const doc = ownerDocument(referenceElement);
   const html = doc.documentElement;
   const body = doc.body;
-  const win = ownerWindow(doc);
-  const htmlStyle = html.style;
-  const bodyStyle = body.style;
-
-  let resizeRaf: number;
-  let scrollX: number;
-  let scrollY: number;
-  let paddingProp: 'paddingLeft' | 'paddingRight';
+  const win = ownerWindow(html);
 
-  function lockScroll() {
-    if (isFirefox) {
-      // RTL <body> scrollbar
-      const scrollbarX =
-        Math.round(doc.documentElement.getBoundingClientRect().left) +
-        doc.documentElement.scrollLeft;
-      paddingProp = scrollbarX ? 'paddingLeft' : 'paddingRight';
-      const scrollbarWidth = win.innerWidth - doc.documentElement.clientWidth;
-
-      bodyStyle.overflow = 'hidden';
-      htmlStyle.overflow = 'visible';
-
-      if (scrollbarWidth) {
-        bodyStyle[paddingProp] = `${scrollbarWidth}px`;
-      }
+  let scrollTop = 0;
+  let scrollLeft = 0;
+  let resizeRaf = -1;
 
-      return;
-    }
+  // Pinch-zoom in Safari causes a shift. Just don't lock scroll if there's any pinch-zoom.
+  if (isWebKit() && (win.visualViewport?.scale ?? 1) !== 1) {
+    return () => {};
+  }
```

Comment:
Should we add a resize handler on the visual viewport to catch when they unzoom?

Reply:
They can still scroll the whole page while pinch-zoomed, so it doesn't really fix the root issue